app/src/main/java/org/onionshare/android/tor/TorManager.kt

```diff
@@ -70,175 +41,107 @@ private const val MOAT_FRONT = "ajax.aspnetcdn.com"
+                } catch (e: Exception) {
+                    LOG.warn("Error while starting Tor: ", e)
+                    stop()
+                }
+            }
+
+            CONNECTED -> LOG.info("connected.")
+            DISABLED -> LOG.info("disabled.")
+        }
     }
 
     @UiThread
     fun onStopped() {
-        if (broadcastReceiverRegistered) {
-            app.unregisterReceiver(broadcastReceiver)
-            app.unregisterReceiver(errorReceiver)
-            broadcastReceiverRegistered = false
-        }
-        localSocket = null
         LOG.info("Stopped")
         val failedToConnect = (state.value as? TorState.Stopping)?.failedToConnect == true
         _state.value = TorState.Stopped(failedToConnect)
     }
 
+    override fun onBootstrapPercentage(percentage: Int) {
+        changeStartingState(percentage)
+    }
+
+    override fun onHsDescriptorUpload(onion: String) {
+        if (state.value !is TorState.Started) changeStartingState(90, onion)
+        onDescriptorUploaded(onion)
+    }
+
+    override fun onClockSkewDetected(skewSeconds: Long) {
+        // TODO
+    }
+
     @Throws(IOException::class, IllegalArgumentException::class)
-    @Suppress("BlockingMethodInNonBlockingContext")
     private suspend fun onTorServiceStarted() = withContext(Dispatchers.IO) {
         changeStartingState(5)
+        val torWrapper = tor ?: return@withContext // Tor was stopped/nulled
         val autoMode = settingsManager.automaticBridges.value
-        controlConnection = startControlConnection().apply {
-            addRawEventListener(onionListener)
-            // create listeners as the first thing to prevent modification while already receiving events
-            launchThread(true)
-            authenticate(ByteArray(0))
-            takeOwnership()
-            setEvents(EVENTS)
-            val obfs4Executable = executableManager.obfs4Executable.absolutePath
-            val snowflakeExecutable = executableManager.snowflakeExecutable.absolutePath
-            val conf = mutableListOf(
-                "ClientTransportPlugin obfs4 exec ${obfs4Executable}$DEBUG_PARAMS_OBFS4",
-                "ClientTransportPlugin meek_lite exec ${obfs4Executable}$DEBUG_PARAMS_OBFS4",
-                "ClientTransportPlugin snowflake exec ${snowflakeExecutable}$DEBUG_PARAMS_SNOWFLAKE",
-            )
-            if (!autoMode) {
-                val customBridges = settingsManager.customBridges.value
-                if (customBridges.isNotEmpty()) {
-                    LOG.info("Using ${customBridges.size} custom bridges...")
-                    conf += listOf("UseBridges 1") + customBridges.map { "Bridge $it" }
-                }
+        if (!autoMode) {
+            val customBridges = settingsManager.customBridges.value
+            if (customBridges.isNotEmpty()) {
+                LOG.info("Using ${customBridges.size} custom bridges...")
+                torWrapper.enableBridges(customBridges)
```

Comment:
Should we defer calling enableNetwork() until we've made this check?

Reply:
Good idea. It seems we get the `CONNECTING` state before enabling the network which seems odd, but works for me.

Reply:
That sounds like a bug to me - I think the wrapper should move from `STARTING_STOPPING` to `DISABLED` until the network is enabled.

Reply:
Created https://code.briarproject.org/briar/onionwrapper/-/issues/10 for this

Reply:
We are now enabling the bridges only after Tor has `STARTED`, then publish hidden service and then enable the network.